File: rustchain-js/src/transaction.js
/**
 * RustChain Transaction Operations
 * 
 * High-level transaction building and sending utilities.
 */

import { TransactionError, AuthenticationError } from './errors.js';

/**
 * High-level transaction operations for RustChain
 * 
 * @example
 * ```javascript
 * const client = new RustChainClient({ adminKey: 'your-key' });
 * const tx = new Transaction(client);
 * 
 * const result = await tx.send('from-wallet', 'to-wallet', 10.0);
 * console.log(`Transaction ID: ${result.pending_id}`);
 * ```
 */
export class Transaction {
  /**
   * Create a Transaction instance
   * @param {Object} client - RustChainClient instance
   */
  constructor(client) {
    this.client = client;
  }
  
  /**
   * Send RTC from one wallet to another
   * 
   * @param {string} fromWallet - Source wallet ID
   * @param {string} toWallet - Destination wallet ID
   * @param {number} amountRtc - Amount to transfer in RTC
   * @param {string} [adminKey] - Optional admin key override
   * @returns {Promise<Object>} Transaction result with pending_id
   * @throws {TransactionError} If transaction fails
   * @throws {AuthenticationError} If admin key is missing
   * 
   * @example
   * ```javascript
   * const result = await tx.send('wallet1', 'wallet2', 10.0);
   * console.log(`Transaction ID: ${result.pending_id}`);
   * ```
   */
  async send(fromWallet, toWallet, amountRtc, adminKey = null) {
    if (amountRtc <= 0) {
      throw new TransactionError('Amount must be greater than 0');
    }
    
    if (!fromWallet || !toWallet) {
      throw new TransactionError('Both fromWallet and toWallet are required');
    }
    
    try {
      const result = await this.client.transferRtc(
        fromWallet,
        toWallet,
        amountRtc,
        adminKey
      );
      
      if (result.error) {
        throw new TransactionError(result.error);
      }
      
      return result;
      
    } catch (error) {
      if (error instanceof AuthenticationError) {
        throw error;
      }
      throw new TransactionError(`Transaction failed: ${error.message}`);
    }
  }
  
  /**
   * Build a transaction without sending it
   * 
   * Useful for previewing transaction details before submission.
   * 
   * @param {string} fromWallet - Source wallet ID
   * @param {string} toWallet - Destination wallet ID
   * @param {number} amountRtc - Amount to transfer in RTC
   * @returns {Object} Transaction preview object
   * 
   * @example
   * ```javascript
   * const preview = tx.buildTransfer('wallet1', 'wallet2', 10.0);
   * console.log(`Will send ${preview.amount_rtc} RTC`);
   * ```
   */
  buildTransfer(fromWallet, toWallet, amountRtc) {
    return {
      from_miner: fromWallet,
      to_miner: toWallet,
      amount_rtc: amountRtc,
      status: 'preview',
    };
  }
  
  /**
   * Validate a transaction before sending
   * 
   * Checks:
   * - Amount is positive
   * - Wallet names are valid format
   * - Source wallet exists (if possible)
   * 
   * @param {string} fromWallet - Source wallet ID
   * @param {string} toWallet - Destination wallet ID
   * @param {number} amountRtc - Amount to transfer in RTC
   * @returns {Promise<[boolean, string]>} Tuple of [is_valid, message]
   * 
   * @example
   * ```javascript
   * const [isValid, msg] = await tx.validateTransfer('wallet1', 'wallet2', 10.0);
   * if (isValid) {
   *   console.log('Transaction is valid!');
   * } else {
   *   console.log(`Invalid: ${msg}`);
   * }
   * ```
   */
  async validateTransfer(fromWallet, toWallet, amountRtc) {
    if (amountRtc <= 0) {
      return [false, 'Amount must be greater than 0'];
    }
    
    if (!fromWallet) {
      return [false, 'Source wallet is required'];
    }
    
    if (!toWallet) {
      return [false, 'Destination wallet is required'];
    }
    
    if (fromWallet === toWallet) {
      return [false, 'Cannot transfer to same wallet'];
    }
    
    // Check if source wallet exists
    const exists = await this.client.checkWalletExists(fromWallet);
    if (!exists) {
      return [false, `Source wallet '${fromWallet}' does not exist`];
    }
    
    return [true, 'Transaction is valid'];
  }
}

File: rustchain-py/rustchain/transaction.py
"""
RustChain Transaction Operations
================================

High-level transaction building and sending utilities.
"""

from typing import Optional, Dict, Any, Tuple
from .client import RustChainClient
from .exceptions import TransactionError, AuthenticationError


class Transaction:
    """
    High-level transaction operations for RustChain.
    
    Provides transaction building, validation, and sending capabilities.
    
    Args:
        client: RustChainClient instance
    
    Example:
        >>> client = RustChainClient(admin_key="your-key")
        >>> tx = Transaction(client)
        >>> result = tx.send("from-wallet", "to-wallet", 10.0)
    """
    
    def __init__(self, client: RustChainClient) -> None:
        """
        Initialize Transaction handler with a RustChain client.
        
        Args:
            client: Authenticated RustChainClient instance for API communication.
                   Should have admin_key set for transaction operations.
        """
        self.client = client
    
    def send(
        self,
        from_wallet: str,
        to_wallet: str,
        amount_rtc: float,
        admin_key: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Send RTC from one wallet to another.
        
        Args:
            from_wallet: Source wallet ID
            to_wallet: Destination wallet ID
            amount_rtc: Amount to transfer in RTC
            admin_key: Optional admin key override
        
        Returns:
            Transaction result dictionary with pending_id
        
        Raises:
            TransactionError: If transaction fails
            AuthenticationError: If admin key is missing
        
        Example:
            >>> result = tx.send("wallet1", "wallet2", 10.0)
            >>> print(f"Transaction ID: {result['pending_id']}")
        """
        if amount_rtc <= 0:
            raise TransactionError("Amount must be greater than 0")
        
        # Validate wallet names
        if not from_wallet or not to_wallet:
            raise TransactionError("Both from_wallet and to_wallet are required")
        
        try:
            result = self.client.transfer_rtc(
                from_wallet=from_wallet,
                to_wallet=to_wallet,
                amount_rtc=amount_rtc,
                admin_key=admin_key
            )
            
            if "error" in result:
                raise TransactionError(result["error"])
            
            return result
            
        except AuthenticationError:
            raise
        except Exception as e:
            raise TransactionError(f"Transaction failed: {str(e)}") from e
    
    def build_transfer(
        self,
        from_wallet: str,
        to_wallet: str,
        amount_rtc: float
    ) -> Dict[str, Any]:
        """
        Build a transaction without sending it.
        
        Useful for previewing transaction details before submission.
        
        Args:
            from_wallet: Source wallet ID
            to_wallet: Destination wallet ID
            amount_rtc: Amount to transfer in RTC
        
        Returns:
            Transaction preview dictionary
        
        Example:
            >>> preview = tx.build_transfer("wallet1", "wallet2", 10.0)
            >>> print(f"Will send {preview['amount_rtc']} RTC")
        """
        return {
            "from_miner": from_wallet,
            "to_miner": to_wallet,
            "amount_rtc": amount_rtc,
            "status": "preview"
        }
    
    def validate_transfer(
        self,
        from_wallet: str,
        to_wallet: str,
        amount_rtc: float
    ) -> Tuple[bool, str]:
        """
        Validate a transaction before sending.
        
        Checks:
            - Amount is positive
            - Wallet names are valid format
            - Source wallet exists (if possible)
        
        Args:
            from_wallet: Source wallet ID
            to_wallet: Destination wallet ID
            amount_rtc: Amount to transfer in RTC
        
        Returns:
            Tuple of (is_valid, message)
        
        Example:
            >>> is_valid, msg = tx.validate_transfer("wallet1", "wallet2", 10.0)
            >>> if is_valid:
            ...     print("Transaction is valid!")
        """
        if amount_rtc <= 0:
            return False, "Amount must be greater than 0"
        
        if not from_wallet:
            return False, "Source wallet is required"
        
        if not to_wallet:
            return False, "Destination wallet is required"
        
        if from_wallet == to_wallet:
            return False, "Cannot transfer to same wallet"
        
        # Check if source wallet exists
        if not self.client.check_wallet_exists(from_wallet):
            return False, f"Source wallet '{from_wallet}' does not exist"
        
        return True, "Transaction is valid"
